File: python/635.design-log-storage-system.py
class LogSystem:
    data = None

    def __init__(self):
        self.data = sortedcontainers.SortedList([])

    def put(self, id: int, timestamp: str) -> None:
        self.data.add([timestamp, id])

    def parse_date(self, timestamp: str, gran:str, replace:chr) -> str:
        res = timestamp
        if gran == "Second":
            return res
        pos = 4
        if gran == "Year": 
            pos = 4
        if gran == "Month":
            pos = 7
        if gran == "Day": 
            pos = 10
        if gran == "Hour": 
            pos = 13
        if gran == "Minute": 
            pos = 16
        return res[:pos]+replace+res[pos+1:]

    def retrieve(self, start: str, end: str, granularity: str) -> List[int]:
        start = self.parse_date(start, granularity, chr(ord(':')-1))
        end = self.parse_date(end, granularity, chr(ord(':')+1))
        left_idx = self.data.bisect_left([start, float('-inf')])
        right_idx = self.data.bisect_right([end, float('inf')])
        return [log[1] for log in self.data[left_idx:right_idx]]


# Your LogSystem object will be instantiated and called as such:
# obj = LogSystem()
# obj.put(id,timestamp)
# param_2 = obj.retrieve(start,end,granularity) param_2 = obj.retrieve(start,end,granularity)

File: cpp/635.design-log-storage-system.cpp
class LogSystem {
public:
    multiset<pair<string, int>> memo;
    LogSystem() {
        
    }
    string parse_date(string &timestamp, string gran, char replace){
	    string res = timestamp;
	    //Year
	    if(gran == "Year") res[4]=replace;
	    //Month
	    if(gran == "Month") res[7]=replace;
	    //Day
	    if(gran == "Day") res[10]=replace;
	    //hour
	    if(gran == "Hour") res[13]=replace;
	    //Minute
	    if(gran == "Minute") res[16]=replace;
	    return res;
    }    
    void put(int id, string timestamp) {
	    memo.insert({timestamp,id});
    }
    
    vector<int> retrieve(string start, string end, string granularity) {
	    vector<int> res;
	    auto l_start = parse_date(start, granularity, ':'-1);
	    auto l_end = parse_date(end, granularity, ':'+1);
	    auto start_it = memo.lower_bound({l_start,-1});
	    auto end_it = memo.upper_bound({l_end, INT_MAX});
	    while(start_it != end_it){
		    res.push_back(start_it->second);
		    start_it++;
	    }
	    return res;
    }
};

/**
 * Your LogSystem object will be instantiated and called as such:
 * LogSystem* obj = new LogSystem();
 * obj->put(id,timestamp);
 * vector<int> param_2 = obj->retrieve(start,end,granularity);
 */
